cmd/postgres_exporter/postgres_exporter.go

```diff
@@ -1425,8 +1425,82 @@ func main() {
 		exporter.servers.Close()
 	}()
 
-	prometheus.MustRegister(exporter)
+	exporter_shared.RunServer("PostgreSQL", *listenAddress, *metricPath, newHandler(map[string]prometheus.Collector{
+		"exporter": exporter,
+	}))
+}
+
+// handler wraps an unfiltered http.Handler but uses a filtered handler,
+// created on the fly, if filtering is requested. Create instances with
+// newHandler. It used for collectors filtering.
+type handler struct {
+	unfilteredHandler http.Handler
+	collectors        map[string]prometheus.Collector
+}
+
+func newHandler(collectors map[string]prometheus.Collector) *handler {
+	h := &handler{collectors: collectors}
+	if innerHandler, err := h.innerHandler(); err != nil {
+		log.Fatalf("Couldn't create metrics handler: %s", err)
+	} else {
+		h.unfilteredHandler = innerHandler
+	}
+	return h
+}
+
+// ServeHTTP implements http.Handler.
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	filters := r.URL.Query()["collect[]"]
+	log.Debugln("collect query:", filters)
+
+	if len(filters) == 0 {
+		// No filters, use the prepared unfiltered handler.
+		h.unfilteredHandler.ServeHTTP(w, r)
+		return
+	}
+
+	filteredHandler, err := h.innerHandler(filters...)
+	if err != nil {
+		log.Warnln("Couldn't create filtered metrics handler:", err)
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(fmt.Sprintf("Couldn't create filtered metrics handler: %s", err)))
+		return
+	}
+	filteredHandler.ServeHTTP(w, r)
+}
+
+func (h *handler) innerHandler(filters ...string) (http.Handler, error) {
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))
+	registry.MustRegister(prometheus.NewGoCollector())
+
+	// register all collectors by default.
+	if len(filters) == 0 {
+		for name, c := range h.collectors {
+			if err := registry.Register(c); err != nil {
+				return nil, err
+			}
+			log.Infof("Collector '%s' was registered", name)
+		}
+	}
+
+	// register only filtered collectors.
+	for _, name := range filters {
+		if c, ok := h.collectors[name]; ok {
+			if err := registry.Register(c); err != nil {
+				return nil, err
+			}
+			log.Infof("Collector '%s' was registered", name)
```

Comment:
We should not log that much by default for every scrape. Use `Debugf`. You may also want to use `%q` instead of `'%s'`.